Check random failures of `cpp11_smart_pointer_compatibility` in async mode

```
checking javascript (napi) testcase cpp11_smart_pointer_compatibility (with run test) (async mode)
file:///home/runner/work/swig/swig/build/build/Examples/test-suite/javascript/cpp11_smart_pointer_compatibility_async_runme.mjs:8
    throw new Error(`expected ${expected}, got ${await resolve(actual)}`);
          ^

Error: expected 51, got -18038784
    at check (file:///home/runner/work/swig/swig/build/build/Examples/test-suite/javascript/cpp11_smart_pointer_compatibility_async_runme.mjs:8:11)

Node.js v18.20.8
make[1]: *** [Makefile:306: cpp11_smart_pointer_compatibility.cpptest] Error 1
```


Probably a race condition or a use-after-free.

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -378,7 +378,7 @@ jobs:
             cxxflags=$($GITHUB_WORKSPACE/Tools/testflags.py --language $SWIGLANG --cxxflags --std=$CPPSTD --compiler=$CC)
             make check-$SWIGLANG-version
             make check-$SWIGLANG-enabled
-            #make $SWIGJOBS check-$SWIGLANG-examples CFLAGS="$cflags" CXXFLAGS="$cxxflags"
+            make $SWIGJOBS check-$SWIGLANG-examples CFLAGS="$cflags" CXXFLAGS="$cxxflags"
             make $SWIGJOBS check-$SWIGLANG-test-suite CC="$TARGET_CC" CXX="$TARGET_CXX" CFLAGS="$cflags" CXXFLAGS="$cxxflags"
           fi
 

--- a/CHANGES-JSE
+++ b/CHANGES-JSE
@@ -8,7 +8,8 @@ Version next
 ===========================
 
 2025-04-23: mmomtchev
-            Integrate `std::shared_ptr` with the JavaScript reference counting of the GC
+            [JavaScript] Integrate `std::shared_ptr` with the JavaScript reference counting
+            of the GC
             [mmomtchev/swig#92](https://github.com/mmomtchev/swig/issues/92), 
             [mmomtchev/swig#95](https://github.com/mmomtchev/swig/issues/95)
 

--- a/CHANGES-JSE
+++ b/CHANGES-JSE
@@ -4,6 +4,14 @@ See the RELEASENOTES file for a summary of changes in each release.
 Issue # numbers mentioned below can be found on Github. For more details, add
 the issue number to the end of the URL: https://github.com/swig/swig/issues/
 
+Version next
+===========================
+
+2025-04-23: mmomtchev
+            Integrate `std::shared_ptr` with the JavaScript reference counting of the GC
+            [mmomtchev/swig#92](https://github.com/mmomtchev/swig/issues/92), 
+            [mmomtchev/swig#95](https://github.com/mmomtchev/swig/issues/95)
+
 Version 5.0.6
 ===========================
 

--- a/Doc/Manual/Javascript.html
+++ b/Doc/Manual/Javascript.html
@@ -39,6 +39,7 @@
 <li><a href="#Javascript_STL_example">STL Containers</a>
 <li><a href="#Javascript_Buffer_example"><code>Buffer</code> and <code>ArrayBuffer</code></a>
 <li><a href="#Javascript_C_arrays_example"><code>C</code> (flat pointer) arrays</a>
+<li><a href="#Javascript_smart_pointer_example">Smart pointers</a>
 </ul>
 <li><a href="#Javascript_exceptions">Exceptions</a>
 <ul>
@@ -881,6 +882,13 @@ This allows for more seamless integration at the price of copying the array at e
   </pre>
 </div>
 <p>By default, these typemaps are available for <code>int</code> and <code>double</code> - check <code>arrays_javascript.i</code> to see how to add more types.</p>
+
+<H3><a name="Javascript_smart_pointer_example">28.4.6 Smart pointers</a></H3>
+
+<p>SWIG for JavaScript supports natively smart pointers with the usual <a href="CPlusPlus11.html#CPlusPlus11_general_purpose_smart_pointers">SWIG semantics</a>. Support for <code>std::shared_ptr</code> is complete, while <code>std::unique_ptr</code> has some caveats since in JavaScript it is impossible to express the fact the a function can take the ownership of a passed object, leaving the original object in an inconsistent state.</p>
+
+<p>Additionally, SWIG Node-API is the only engine which provides an integration of the reference counting of a <code>std::shared_ptr</code> with the reference counting of the GC. This means that Node-API allows C++ to keep a copy of a passed <code>std::shared_ptr</code> and the GC won't destroy this object until C++ has deleted its last copy.</p>
+
 <H2><a name="Javascript_exceptions">28.5 Exceptions</a></H2>
 
 <H3><a name="Javascript_jsc_exceptions">28.5.1 Handling Exceptions in JavascriptCore</a></H3>

--- a/Examples/test-suite/javascript/cpp11_smart_pointer_compatibility_runme.js
+++ b/Examples/test-suite/javascript/cpp11_smart_pointer_compatibility_runme.js
@@ -24,45 +24,45 @@ var sp42 = /* await */(cpp11_smart_pointer_compatibility.ProduceSharedPointer(42
 var spp77 = /* await */(cpp11_smart_pointer_compatibility.ProduceSharedPointerRef(77));
 var ar51 = /* await */(cpp11_smart_pointer_compatibility.ProduceUniquePointer(51));
 
-check(i17.get(), 17);
-check(sp42.get(), 42);
-check(spp77.get(), 77);
-check(ar51.get(), 51);
+/* await */(check(i17.get(), 17));
+/* await */(check(sp42.get(), 42));
+/* await */(check(spp77.get(), 77));
+/* await */(check(ar51.get(), 51));
 
-check(cpp11_smart_pointer_compatibility.ConsumePlainObject(i17), 17);
-check(cpp11_smart_pointer_compatibility.ConsumePlainObject(sp42), 42);
-check(cpp11_smart_pointer_compatibility.ConsumePlainObject(spp77), 77);
-check(cpp11_smart_pointer_compatibility.ConsumePlainObject(ar51), 51);
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePlainObject(i17), 17));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePlainObject(sp42), 42));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePlainObject(spp77), 77));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePlainObject(ar51), 51));
 
-check(cpp11_smart_pointer_compatibility.ConsumeConstReference(i17), 17);
-check(cpp11_smart_pointer_compatibility.ConsumeConstReference(sp42), 42);
-check(cpp11_smart_pointer_compatibility.ConsumeConstReference(spp77), 77);
-check(cpp11_smart_pointer_compatibility.ConsumeConstReference(ar51), 51);
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeConstReference(i17), 17));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeConstReference(sp42), 42));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeConstReference(spp77), 77));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeConstReference(ar51), 51));
 
-check(cpp11_smart_pointer_compatibility.ConsumePointer(i17), 17);
-check(cpp11_smart_pointer_compatibility.ConsumePointer(sp42), 42);
-check(cpp11_smart_pointer_compatibility.ConsumePointer(spp77), 77);
-check(cpp11_smart_pointer_compatibility.ConsumePointer(ar51), 51);
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePointer(i17), 17));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePointer(sp42), 42));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePointer(spp77), 77));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumePointer(ar51), 51));
 
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(i17), 17);
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(sp42), 42);
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(spp77), 77);
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(ar51), 51);
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(i17), 17));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(sp42), 42));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(spp77), 77));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointer(ar51), 51));
 
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(i17), 17);
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(sp42), 42);
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(spp77), 77);
-check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(ar51), 51);
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(i17), 17));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(sp42), 42));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(spp77), 77));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeSharedPointerReference(ar51), 51));
 
 
-check(cpp11_smart_pointer_compatibility.ConsumeUniquePtr(i17), 17);
-check(cpp11_smart_pointer_compatibility.ConsumeUniquePtr(ar51), 51);
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeUniquePtr(i17), 17));
+/* await */(check(cpp11_smart_pointer_compatibility.ConsumeUniquePtr(ar51), 51));
 
 // These functions expect unique_ptr&& and take ownership
-checkThrows(/* async */() => {
+/* await */(checkThrows(/* async */() => {
   /* await */(cpp11_smart_pointer_compatibility.ConsumeUniquePtr(sp42));
-}, /memory is not owned/);
-checkThrows(/* async */() => {
+}, /memory is not owned/));
+/* await */(checkThrows(/* async */() => {
   /* await */(cpp11_smart_pointer_compatibility.ConsumeUniquePtr(spp77));
-}, /memory is not owned/);
+}, /memory is not owned/));
 

--- a/Lib/javascript/napi/javascriptinit.swg
+++ b/Lib/javascript/napi/javascriptinit.swg
@@ -30,9 +30,8 @@ EnvInstanceData::EnvInstanceData(Napi::Env env, swig_module_info *swig_module) :
   js_main_thread.handle = new uv_async_t;
   uv_loop_t *event_loop;
   napi_status r = napi_get_uv_event_loop(env, &event_loop);
-  if (r != napi_ok) throw Napi::Error::New(env, "Failed retrieving libuv event loop");
-  if (uv_async_init(event_loop, js_main_thread.handle, SWIG_NAPI_RunMainThreadQueue) != 0)
-    throw Napi::Error::New(env, "Failed initializing libuv queue");
+  assert(r == napi_ok);
+  assert(uv_async_init(event_loop, js_main_thread.handle, SWIG_NAPI_RunMainThreadQueue) == 0);
   // The initial status of the async handle is inactive
 #ifndef __EMSCRIPTEN__
   uv_unref(reinterpret_cast<uv_handle_t *>(js_main_thread.handle));

--- a/Lib/javascript/napi/javascriptinit.swg
+++ b/Lib/javascript/napi/javascriptinit.swg
@@ -24,6 +24,7 @@ EnvInstanceData::EnvInstanceData(Napi::Env env, swig_module_info *swig_module) :
     ctor[i] = SWIG_NULLPTR;
   }
 
+#if NAPI_HAS_THREADS
   // Init the mechanism for scheduling jobs on the JS main thread
   js_main_thread.main_thread_id = std::this_thread::get_id();
   js_main_thread.handle = new uv_async_t;
@@ -35,6 +36,7 @@ EnvInstanceData::EnvInstanceData(Napi::Env env, swig_module_info *swig_module) :
   // The initial status of the async handle is inactive
   uv_unref(reinterpret_cast<uv_handle_t *>(js_main_thread.handle));
   js_main_thread.handle->data = static_cast<void*>(this);
+#endif
 }
 
 EnvInstanceData::~EnvInstanceData() {

--- a/Lib/javascript/napi/javascriptinit.swg
+++ b/Lib/javascript/napi/javascriptinit.swg
@@ -34,7 +34,9 @@ EnvInstanceData::EnvInstanceData(Napi::Env env, swig_module_info *swig_module) :
   if (uv_async_init(event_loop, js_main_thread.handle, SWIG_NAPI_RunMainThreadQueue) != 0)
     throw Napi::Error::New(env, "Failed initializing libuv queue");
   // The initial status of the async handle is inactive
+#ifndef __EMSCRIPTEN__
   uv_unref(reinterpret_cast<uv_handle_t *>(js_main_thread.handle));
+#endif
   js_main_thread.handle->data = static_cast<void*>(this);
 #endif
 }

--- a/Lib/javascript/napi/javascriptinit.swg
+++ b/Lib/javascript/napi/javascriptinit.swg
@@ -23,6 +23,18 @@ EnvInstanceData::EnvInstanceData(Napi::Env env, swig_module_info *swig_module) :
   for (size_t i = 0; i <= swig_module->size; i++) {
     ctor[i] = SWIG_NULLPTR;
   }
+
+  // Init the mechanism for scheduling jobs on the JS main thread
+  js_main_thread.main_thread_id = std::this_thread::get_id();
+  js_main_thread.handle = new uv_async_t;
+  uv_loop_t *event_loop;
+  napi_status r = napi_get_uv_event_loop(env, &event_loop);
+  if (r != napi_ok) throw Napi::Error::New(env, "Failed retrieving libuv event loop");
+  if (uv_async_init(event_loop, js_main_thread.handle, SWIG_NAPI_RunMainThreadQueue) != 0)
+    throw Napi::Error::New(env, "Failed initializing libuv queue");
+  // The initial status of the async handle is inactive
+  uv_unref(reinterpret_cast<uv_handle_t *>(js_main_thread.handle));
+  js_main_thread.handle->data = static_cast<void*>(this);
 }
 
 EnvInstanceData::~EnvInstanceData() {

--- a/Lib/javascript/napi/javascriptrun.swg
+++ b/Lib/javascript/napi/javascriptrun.swg
@@ -2,10 +2,12 @@
  * SWIG Node-API runtime : declarations
  * ---------------------------------------------------------------------------*/
 
+#if NAPI_HAS_THREADS
 #include <thread>
 #include <queue>
 #include <mutex>
 #include <uv.h>
+#endif
 
 /*
  * We support several forms:
@@ -126,6 +128,7 @@ struct EnvInstanceData {
   Napi::FunctionReference **ctor;
   swig_module_info *swig_module;
   void *module_instance_data;
+#if NAPI_HAS_THREADS
   // Queue for running jobs on the JS main thread from
   // background threads
   struct {
@@ -134,14 +137,17 @@ struct EnvInstanceData {
     std::queue<std::function<void()>> jobs;
     std::mutex lock;
   } js_main_thread;
+#endif
   EnvInstanceData(Napi::Env, swig_module_info *);
   ~EnvInstanceData();
 };
 
 // Execute the jobs on the queue on the main thread
 void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *, std::function<void()> &&);
+#if NAPI_HAS_THREADS
 // Schedule a job to run on the main thread
 void SWIG_NAPI_RunMainThreadQueue(uv_async_t *);
+#endif
 
 typedef size_t SWIG_NAPI_ClientData;
 

--- a/Lib/javascript/napi/javascriptrun.swg
+++ b/Lib/javascript/napi/javascriptrun.swg
@@ -2,6 +2,11 @@
  * SWIG Node-API runtime : declarations
  * ---------------------------------------------------------------------------*/
 
+#include <thread>
+#include <queue>
+#include <mutex>
+#include <uv.h>
+
 /*
  * We support several forms:
  *
@@ -121,9 +126,22 @@ struct EnvInstanceData {
   Napi::FunctionReference **ctor;
   swig_module_info *swig_module;
   void *module_instance_data;
+  // Queue for running jobs on the JS main thread from
+  // background threads
+  struct {
+    uv_async_t *handle;
+    std::thread::id main_thread_id;
+    std::queue<std::function<void()>> jobs;
+    std::mutex lock;
+  } js_main_thread;
   EnvInstanceData(Napi::Env, swig_module_info *);
   ~EnvInstanceData();
 };
+
+// Execute the jobs on the queue on the main thread
+void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *, std::function<void()> &&);
+// Schedule a job to run on the main thread
+void SWIG_NAPI_RunMainThreadQueue(uv_async_t *);
 
 typedef size_t SWIG_NAPI_ClientData;
 

--- a/Lib/javascript/napi/javascriptrunbody.swg
+++ b/Lib/javascript/napi/javascriptrunbody.swg
@@ -397,6 +397,7 @@ Napi::Value SWIG_NAPI_AppendOutput(Napi::Env env, Napi::Value result, Napi::Valu
 /* ---------------------------------------------------------------------------
  * Execute the jobs on the queue on the main thread
  * ---------------------------------------------------------------------------*/
+#if NAPI_HAS_THREADS
 void SWIG_NAPI_RunMainThreadQueue(uv_async_t *async) {
   printf("Running on the main thread\n");
   auto env_data = reinterpret_cast<EnvInstanceData *>(async->data);
@@ -411,12 +412,14 @@ void SWIG_NAPI_RunMainThreadQueue(uv_async_t *async) {
   // (we do not want to block Node from exiting)
   uv_unref(reinterpret_cast<uv_handle_t *>(env_data->js_main_thread.handle));
 }
+#endif
 
 
 /* ---------------------------------------------------------------------------
  * Schedule a job to run on the main thread
  * ---------------------------------------------------------------------------*/
 void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *env_data, std::function<void()> &&job) {
+#if NAPI_HAS_THREADS
   if (env_data->js_main_thread.main_thread_id == std::this_thread::get_id()) {
     printf("Synchronously running on the main thread\n");
     job();
@@ -427,4 +430,8 @@ void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *env_data, std::function<void()
     assert(uv_async_send(env_data->js_main_thread.handle) == 0);
     uv_ref(reinterpret_cast<uv_handle_t *>(env_data->js_main_thread.handle));
   }
+#else
+  printf("Running on the main thread w/o threads\n");
+  job();
+#endif
 }

--- a/Lib/javascript/napi/javascriptrunbody.swg
+++ b/Lib/javascript/napi/javascriptrunbody.swg
@@ -410,7 +410,9 @@ void SWIG_NAPI_RunMainThreadQueue(uv_async_t *async) {
   }
   // Disable the async because the queue is empty
   // (we do not want to block Node from exiting)
+#ifndef __EMSCRIPTEN__
   uv_unref(reinterpret_cast<uv_handle_t *>(env_data->js_main_thread.handle));
+#endif
 }
 #endif
 
@@ -428,7 +430,9 @@ void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *env_data, std::function<void()
     std::lock_guard<std::mutex> lock(env_data->js_main_thread.lock);
     env_data->js_main_thread.jobs.emplace(std::move(job));
     assert(uv_async_send(env_data->js_main_thread.handle) == 0);
+#ifndef __EMSCRIPTEN__
     uv_ref(reinterpret_cast<uv_handle_t *>(env_data->js_main_thread.handle));
+#endif
   }
 #else
   printf("Running on the main thread w/o threads\n");

--- a/Lib/javascript/napi/javascriptrunbody.swg
+++ b/Lib/javascript/napi/javascriptrunbody.swg
@@ -399,7 +399,9 @@ Napi::Value SWIG_NAPI_AppendOutput(Napi::Env env, Napi::Value result, Napi::Valu
  * ---------------------------------------------------------------------------*/
 #if NAPI_HAS_THREADS
 void SWIG_NAPI_RunMainThreadQueue(uv_async_t *async) {
-  printf("Running on the main thread\n");
+#ifdef SWIGRUNTIME_DEBUG
+  printf("Running scheduled tasks on main thread\n");
+#endif
   auto env_data = reinterpret_cast<EnvInstanceData *>(async->data);
 
   // As the lambdas are very light, it is better to not release the lock at all
@@ -423,10 +425,14 @@ void SWIG_NAPI_RunMainThreadQueue(uv_async_t *async) {
 void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *env_data, std::function<void()> &&job) {
 #if NAPI_HAS_THREADS
   if (env_data->js_main_thread.main_thread_id == std::this_thread::get_id()) {
-    printf("Synchronously running on the main thread\n");
+#ifdef SWIGRUNTIME_DEBUG
+    printf("Running scheduled task synchronously on the main thread\n");
+#endif
     job();
   } else {
-    printf("Scheduling on the main thread\n");
+#ifdef SWIGRUNTIME_DEBUG
+    printf("Scheduling a task on the main thread\n");
+#endif
     std::lock_guard<std::mutex> lock(env_data->js_main_thread.lock);
     env_data->js_main_thread.jobs.emplace(std::move(job));
     assert(uv_async_send(env_data->js_main_thread.handle) == 0);
@@ -435,7 +441,9 @@ void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *env_data, std::function<void()
 #endif
   }
 #else
-  printf("Running on the main thread w/o threads\n");
+#ifdef SWIGRUNTIME_DEBUG
+  printf("Running a task on the main thread w/o thread support\n");
+#endif
   job();
 #endif
 }

--- a/Lib/javascript/napi/javascriptrunbody.swg
+++ b/Lib/javascript/napi/javascriptrunbody.swg
@@ -393,3 +393,38 @@ Napi::Value SWIG_NAPI_AppendOutput(Napi::Env env, Napi::Value result, Napi::Valu
   arr.Set(arr.Length(), obj);
   return arr;
 }
+
+/* ---------------------------------------------------------------------------
+ * Execute the jobs on the queue on the main thread
+ * ---------------------------------------------------------------------------*/
+void SWIG_NAPI_RunMainThreadQueue(uv_async_t *async) {
+  printf("Running on the main thread\n");
+  auto env_data = reinterpret_cast<EnvInstanceData *>(async->data);
+
+  // As the lambdas are very light, it is better to not release the lock at all
+  std::lock_guard<std::mutex> lock(env_data->js_main_thread.lock);
+  while (!env_data->js_main_thread.jobs.empty()) {
+    env_data->js_main_thread.jobs.front()();
+    env_data->js_main_thread.jobs.pop();
+  }
+  // Disable the async because the queue is empty
+  // (we do not want to block Node from exiting)
+  uv_unref(reinterpret_cast<uv_handle_t *>(env_data->js_main_thread.handle));
+}
+
+
+/* ---------------------------------------------------------------------------
+ * Schedule a job to run on the main thread
+ * ---------------------------------------------------------------------------*/
+void SWIG_NAPI_RunOnJSMainThread(EnvInstanceData *env_data, std::function<void()> &&job) {
+  if (env_data->js_main_thread.main_thread_id == std::this_thread::get_id()) {
+    printf("Synchronously running on the main thread\n");
+    job();
+  } else {
+    printf("Scheduling on the main thread\n");
+    std::lock_guard<std::mutex> lock(env_data->js_main_thread.lock);
+    env_data->js_main_thread.jobs.emplace(std::move(job));
+    assert(uv_async_send(env_data->js_main_thread.handle) == 0);
+    uv_ref(reinterpret_cast<uv_handle_t *>(env_data->js_main_thread.handle));
+  }
+}

--- a/Lib/javascript/napi/std_shared_ptr.i
+++ b/Lib/javascript/napi/std_shared_ptr.i
@@ -37,7 +37,9 @@
   auto env_data = env.GetInstanceData<EnvInstanceData>();
   $1 = std::shared_ptr<CONST TYPE>(plain_ptr, [persistent, env_data](void *){
     SWIG_NAPI_RunOnJSMainThread(env_data, [persistent](){
-      printf("Releasing pointer\n");
+#ifdef SWIGRUNTIME_DEBUG
+      printf("Releasing persistent reference for shared_ptr\n");
+#endif
       delete persistent;
     });
   });
@@ -56,7 +58,9 @@
   auto env_data = env.GetInstanceData<EnvInstanceData>();
   $1 = new std::shared_ptr<CONST TYPE>(plain_ptr, [persistent, env_data](void *){
     SWIG_NAPI_RunOnJSMainThread(env_data, [persistent](){
-      printf("Releasing pointer\n");
+#ifdef SWIGRUNTIME_DEBUG
+      printf("Releasing persistent reference for shared_ptr\n");
+#endif
       delete persistent;
     });
   });

--- a/Lib/javascript/napi/std_shared_ptr.i
+++ b/Lib/javascript/napi/std_shared_ptr.i
@@ -9,6 +9,11 @@
  * - to use it in place of a plain object as a function input argument 
  * - to use a plain object where a shared pointer is expected
  *
+ * Additionally, shared_ptr to JS created objects passed to C++
+ * are counted as references to the JS objects protecting them
+ * from the GC. This means that C++ can keep a copy of a passed
+ * shared_ptr.
+ *
  * ----------------------------------------------------------------------------- */
 
 #define SWIG_SHARED_PTR_NAMESPACE std
@@ -16,22 +21,33 @@
 
 %define SWIG_SHARED_PTR_TYPEMAPS(CONST, TYPE)
 
-%typemap(in, fragment="SWIG_null_deleter") std::shared_ptr<CONST TYPE> {
+// These typemaps obtain a JS persistent reference
+// that is freed when the shared_ptr is destroyed,
+// signaling the GC that C++ has released the value
+%typemap(in) std::shared_ptr<CONST TYPE> {
   TYPE *plain_ptr;
   int res = SWIG_ConvertPtr($input, reinterpret_cast<void**>(&plain_ptr), $descriptor(TYPE *), %convertptr_flags);
   if (!SWIG_IsOK(res)) {
     %argument_fail(res, "TYPE", $symname, $argnum);
   }
-  $1 = std::shared_ptr<CONST TYPE>(plain_ptr, SWIG_null_deleter());
+  Napi::Reference<Napi::Value> *persistent = new Napi::Reference<Napi::Value>;
+  *persistent = Napi::Persistent($input);
+  $1 = std::shared_ptr<CONST TYPE>(plain_ptr, [persistent](void *){
+    delete persistent;
+  });
 }
 
-%typemap(in, fragment="SWIG_null_deleter") std::shared_ptr<CONST TYPE> *, std::shared_ptr<CONST TYPE> & {
+%typemap(in) std::shared_ptr<CONST TYPE> *, std::shared_ptr<CONST TYPE> & {
   TYPE *plain_ptr;
   int res = SWIG_ConvertPtr($input, reinterpret_cast<void**>(&plain_ptr), $descriptor(TYPE *), %convertptr_flags);
   if (!SWIG_IsOK(res)) {
     %argument_fail(res, "TYPE", $symname, $argnum);
   }
-  $1 = new std::shared_ptr<CONST TYPE>(plain_ptr, SWIG_null_deleter());
+  Napi::Reference<Napi::Value> *persistent = new Napi::Reference<Napi::Value>;
+  *persistent = Napi::Persistent($input);
+  $1 = new std::shared_ptr<CONST TYPE>(plain_ptr, [persistent](void *){
+    delete persistent;
+  });
 }
 %typemap(freearg) std::shared_ptr<CONST TYPE> *, std::shared_ptr<CONST TYPE> & {
   delete $1;

--- a/Lib/javascript/napi/std_shared_ptr.i
+++ b/Lib/javascript/napi/std_shared_ptr.i
@@ -34,8 +34,12 @@
   if (!$input.IsNull()) {
     *persistent = Napi::Persistent($input);
   }
-  $1 = std::shared_ptr<CONST TYPE>(plain_ptr, [persistent](void *){
-    delete persistent;
+  auto env_data = env.GetInstanceData<EnvInstanceData>();
+  $1 = std::shared_ptr<CONST TYPE>(plain_ptr, [persistent, env_data](void *){
+    SWIG_NAPI_RunOnJSMainThread(env_data, [persistent](){
+      printf("Releasing pointer\n");
+      delete persistent;
+    });
   });
 }
 
@@ -49,8 +53,12 @@
   if (!$input.IsNull()) {
     *persistent = Napi::Persistent($input);
   }
-  $1 = new std::shared_ptr<CONST TYPE>(plain_ptr, [persistent](void *){
-    delete persistent;
+  auto env_data = env.GetInstanceData<EnvInstanceData>();
+  $1 = new std::shared_ptr<CONST TYPE>(plain_ptr, [persistent, env_data](void *){
+    SWIG_NAPI_RunOnJSMainThread(env_data, [persistent](){
+      printf("Releasing pointer\n");
+      delete persistent;
+    });
   });
 }
 %typemap(freearg) std::shared_ptr<CONST TYPE> *, std::shared_ptr<CONST TYPE> & {

--- a/Lib/javascript/napi/std_shared_ptr.i
+++ b/Lib/javascript/napi/std_shared_ptr.i
@@ -31,7 +31,9 @@
     %argument_fail(res, "TYPE", $symname, $argnum);
   }
   Napi::Reference<Napi::Value> *persistent = new Napi::Reference<Napi::Value>;
-  *persistent = Napi::Persistent($input);
+  if (!$input.IsNull()) {
+    *persistent = Napi::Persistent($input);
+  }
   $1 = std::shared_ptr<CONST TYPE>(plain_ptr, [persistent](void *){
     delete persistent;
   });
@@ -44,7 +46,9 @@
     %argument_fail(res, "TYPE", $symname, $argnum);
   }
   Napi::Reference<Napi::Value> *persistent = new Napi::Reference<Napi::Value>;
-  *persistent = Napi::Persistent($input);
+  if (!$input.IsNull()) {
+    *persistent = Napi::Persistent($input);
+  }
   $1 = new std::shared_ptr<CONST TYPE>(plain_ptr, [persistent](void *){
     delete persistent;
   });

